js/jquery.mapael.js

```diff
@@ -630,6 +635,164 @@
+            if (opt.ranges && opt.ranges.plot) {
+                self.showElemByRange(opt.ranges.plot, self.plots, opt.hiddenOpacity, opt.animDuration);
+            }
+
+            // handle link
+            if (opt.ranges && opt.ranges.link) {
+                self.showElemByRange(opt.ranges.link, self.links, opt.hiddenOpacity, opt.animDuration);
+            }
+
+            // Call user callback
+            if (opt.afterShowRange) opt.afterShowRange();
+        },
+
+        /*
+         * Show some element in range
+         * @param ranges: the ranges
+         * @param elems: list of element on which to check against previous range
+         * @hiddenOpacity: the opacity when hidden
+         * @animDuration: the animation duration
+         */
+        showElemByRange: function(ranges, elems, hiddenOpacity, animDuration) {
+            var self = this;
+            // Hold the final opacity value for all elements consolidated after applying each ranges
+            // This allow to set the opacity only once for each elements
+            var elemsFinalOpacity = {};
+
+            // set object with one valueIndex to 0 if we have directly the min/max
+            if (ranges.min !== undefined || ranges.max !== undefined) {
+                ranges = {0: ranges};
+            }
+
+            // Loop through each valueIndex
+            $.each(ranges, function (valueIndex) {
+                var range = ranges[valueIndex];
+                // Check if user defined at least a min or max value
+                if (range.min === undefined && range.max === undefined) {
+                    return true; // next range
+                }
+                // Loop through each elements
+                $.each(elems, function (id) {
+                    var elemValue = elems[id].value;
+                    // set value with one valueIndex to 0 if not object
+                    if (typeof elemValue !== "object") {
+                        elemValue = [elemValue];
+                    }
+                    // Check existence of this value index
+                    if (elemValue[valueIndex] === undefined) {
+                        return true; // next element
+                    }
+                    // Check if in range
+                    if ((range.min !== undefined && elemValue[valueIndex] < range.min) ||
+                        (range.max !== undefined && elemValue[valueIndex] > range.max)) {
+                        // Element not in range
+                        elemsFinalOpacity[id] = hiddenOpacity;
+                    } else {
+                        // Element in range
+                        elemsFinalOpacity[id] = 1;
+                    }
+                });
+            });
+            // Now that we looped through all ranges, we can really assign the final opacity
+            $.each(elemsFinalOpacity, function (id) {
+                self.setElementOpacity(elems[id], elemsFinalOpacity[id], animDuration);
```

Comment:
I made two loops on purpose:
- the first go through each ranges, and then checks each element if in range, and define an opacity to this element
- the second loop will take this opacity and apply it.

Hence, if an element is in a range first range (opacity = 1), but not in a second range (opacity = 0), the element opacity will be set to 1 then 0. With `animate()`, this could cause flickering. Thus why I first check all ranges, and apply only once a *final opacity*.